technical/binary_tree/c++/binary_tree.h

```diff
@@ -0,0 +1,272 @@
+  return root == nullptr;
+}
+
+template <typename T> 
+typename binary_tree<T>::iterator binary_tree<T>::first() {
+  iterator i;
+  node* first = root;
+  while(first) {
+    i.nodes.push(first);
+    if (first->left) {
+      first = first->left;
+    }
+    first = first->right;
+  }
+  return i;
+}
+
+template <typename T>
+typename binary_tree<T>::iterator binary_tree<T>::end() {
+  return iterator();
+}
+
+template <typename T>
+void binary_tree<T>::print_helper(node* n) {
+  if(n) {
+    print_helper(n->left);
+    std::cout << n->data << " ";
+    print_helper(n->right);
+  }
+}
+
+template <typename T>
+void binary_tree<T>::print() {
+  std::cout << "[ ";
+  print_helper(root);
+  std::cout << "]\n";
+}
+
+template <typename T>
+typename binary_tree<T>::node*& binary_tree<T>::find_node(T val, binary_tree<T>::node*& n) {
+  if(!n) return n;
+  else if(n->data < val) {
+    return find_node(val, n->right);
+  } else if (val < n->data) {
+    return find_node(val, n->left);
+  } else return n;
+}
+
+template <typename T>
+typename binary_tree<T>::node* binary_tree<T>::in_order_successor(binary_tree<T>::node* n) {
+  if(!n->right) return nullptr;
+  binary_tree<T>::node* candidate = n->right;
+  while(candidate->left) {
+    candidate = candidate->left;
+  }
+  return candidate;
+} 
+
+template <typename T>
+typename binary_tree<T>::node* binary_tree<T>::in_order_predecessor(binary_tree<T>::node* n){
+  if (!n->left) return nullptr;
+  binary_tree<T>::node* candidate = n->left;
+  while(candidate->right) {
+    candidate = candidate->right;
+  }
+  return candidate;
+}
+
+template <typename T>
+void binary_tree<T>::remove_helper(binary_tree<T>::node* & n) {
+  if(!n) return;
+  binary_tree<T>::node* successor = in_order_successor(n);
+  if(successor) {
+    n->data = successor->data;
+    remove_helper(successor);
+    return;
+  }
+  binary_tree<T>::node* predecessor = in_order_predecessor(n);
+  if(predecessor) {
+    n->data = predecessor->data;
+    remove_helper(predecessor);
+    return;
+  }
+  // n is a leaf node. Just delete.
+  binary_tree<T>::node* parent = n->parent;
+  if(!parent) {
+    if(n == root) {
+      delete root;
+      root = nullptr;
+    } else {
+      delete n;
+      n = nullptr;
+    }
+  } else if(parent->left == n) {
+    delete parent->left;
+    parent->left = nullptr;
+  } else {
+    delete parent->right;
+    parent->right = nullptr; 
+  }
+}
+
+template <typename T>
+void binary_tree<T>::remove(T val) {
+  binary_tree<T>::node* to_delete = find_node(val, root);
+  if(!to_delete) return;
+  remove_helper(to_delete);
+}
+
+
+
+/// ITERATOR CLASS IMPLEMENTATION
+
+
+template <typename T>
+binary_tree<T>::iterator::iterator() {}
+
+template <typename T>
+binary_tree<T>::iterator::iterator(const iterator& other) {
+  nodes = other.nodes;
+}
+
+template <typename T>
+typename binary_tree<T>::iterator binary_tree<T>::iterator::operator=(const iterator& other) {
+  this->nodes = other.nodes;
```

Comment:
I can't remember if it was special with templates, but I'm pretty sure the use of "this" is redundant. Prefer to omit "this" when possible